lib/OpenQA/Scheduler/Scheduler.pm

```diff
@@ -138,17 +150,397 @@ sub _prefer_parallel {
+                  and return ()
+                  if @free_workers == 0
+                  && OpenQA::Scheduler::CONGESTION_CONTROL()
+                  && OpenQA::Scheduler::BUSY_BACKOFF()
+                  && $all_workers > 0;
+                my $allocating = {};
+                for my $w (@free_workers) {
+
+                    my @possible_jobs = job_grab(
+                        workerid     => $w->id(),
+                        blocking     => 0,
+                        allocate     => 0,
+                        scheduler    => 1,
+                        jobs         => OpenQA::Scheduler::MAX_JOB_ALLOCATION(),
+                        max_attempts => OpenQA::Scheduler::FIND_JOB_ATTEMPTS());
+                    log_debug("[Worker#" . $w->id() . "] Possible jobs to allocate: " . scalar(@possible_jobs));
+                    my $allocated_job;
+                    for my $p_job (@possible_jobs) {
+                        next if exists $p_job->{id} && exists $allocating->{$p_job->{id}};
+                        last if $allocated_job;
+                        $allocated_job = $p_job;
+                    }
+                    log_debug(pp(\@possible_jobs));
+                    log_debug("[Worker#" . $w->id() . "] Among them, we have chosen job: " . $allocated_job->{id})
+                      if $allocated_job && exists $allocated_job->{id};
+
+                    $allocated_job->{assigned_worker_id} = $w->id() if $allocated_job;
+                    next unless $allocated_job && exists $allocated_job->{id};
+
+
+      # TODO: we need to be sure job_grab is not returning the same job for different workers - for now do not push them
+                    push(@allocated_jobs, $allocated_job) if !exists $allocating->{$allocated_job->{id}};
+                    $allocating->{$allocated_job->{id}}++;
+                }
+                return @allocated_jobs;
+            });
+    }
+    catch {
+        # we had a real failure
+        $failure++ if OpenQA::Scheduler::CONGESTION_CONTROL();
+    };
+
+    @allocated_jobs = splice @allocated_jobs, 0, OpenQA::Scheduler::MAX_JOB_ALLOCATION()
+      if (OpenQA::Scheduler::MAX_JOB_ALLOCATION() > 0
+        && scalar(@allocated_jobs) > OpenQA::Scheduler::MAX_JOB_ALLOCATION());
+
+    my $successfully_allocated = 0;
+
+    foreach my $allocated (@allocated_jobs) {
+        #  Now we need to set  the worker in the job, with the state in SCHEDULED.
+
+        my $job = schema->resultset("Jobs")->find({id => $allocated->{id}});
+        my $worker = schema->resultset("Workers")->find({id => $allocated->{assigned_worker_id}});
+        my $res;
+        try {
+            $res = $job->ws_send($worker);
+        }
+        catch {
+            log_debug("Failed to send data to websocket :( bummer! Reason: $_");
+            $failure++
+              if OpenQA::Scheduler::CONGESTION_CONTROL()
+              && OpenQA::Scheduler::BUSY_BACKOFF();    # double it, we might be in a heavy load condition.
```

Comment:
these two conditions are appearing together often enough. I recommend to combine into one condition with good name, don't you think?